api/v3/Payment.php

```diff
@@ -0,0 +1,291 @@
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to refund a payment'));
+  }
+  $eftParams = array(
+    'entity_table' => 'civicrm_contribution',
+    'financial_trxn_id' => $ftId,
+  );
+  $entity = civicrm_api3('EntityFinancialTrxn', 'get', $eftParams);
+  $entity = reset($entity['values']);
+  $contributionId = $entity['entity_id'];
+  $params['total_amount'] = $entity['amount'];
+  
+  $trxn = CRM_Contribute_BAO_Contribution::recordAdditionalPayment($contributionId, $params, 'refund', NULL, FALSE);
+  
+  $values = array(); 
+  if (is_a($trxn, 'CRM_Core_Error')) {
+    throw new API_Exception($trxn->_errors[0]['message']);
+  }
+  else {
+    _civicrm_api3_object_to_array_unique_fields($trxn, $values[$trxn->id]);
+  }
+  return civicrm_api3_create_success($values, $params, 'Payment', 'cancel', $trxn);
+}
+
+/**
+ * Add a payment for a Contribution.
+ *
+ * @param array $params
+ *   Input parameters.
+ *
+ * @throws API_Exception
+ * @return array
+ *   Api result array
+ */
+function civicrm_api3_payment_create(&$params) {
+  if (empty($params['contribution_id']) || 
+      (isset($params['contribution_id']) && !CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $params['contribution_id'], 'id'))) {
+    return civicrm_api3_create_error(ts('You need to supply a valid contribution ID to create a payment'));
+  }
+  // Check if it is an update
+  $ftId = isset($params['civicrm_financial_trxn_id']) ? $params['civicrm_financial_trxn_id'] : $params['payment_id'];
+  if ($ftId && (!CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'id') || !CRM_Core_DAO::getFieldValue('CRM_Financial_DAO_FinancialTrxn', $ftId, 'is_payment'))) {
+    return civicrm_api3_create_error(ts('You need to supply a valid Payment ID to update a payment'));
+  }
+  elseif ($ftId) {
+    $amount = $params['total_amount'];
+    civicrm_api3('Payment', 'cancel', $params);
+    $params['total_amount'] = $amount;
+  }
+  // Get contribution
+  $contribution = civicrm_api3('Contribution', 'get', array('id' => $params['contribution_id']));
+  $contribution = reset($contribution['values']);
+  if ($contribution['contribution_status'] != 'Partially paid') {
+    return civicrm_api3_create_error(ts('Please select a contribution which has a partial payment'));
+  }
+  else {
+    $trxn = CRM_Contribute_BAO_Contribution::recordPartialPayment($contribution, $params);
+    $paid = CRM_Core_BAO_FinancialTrxn::getTotalPayments($params['contribution_id']);
+    $total = CRM_Core_DAO::getFieldValue('CRM_Contribute_DAO_Contribution', $params['contribution_id'], 'total_amount');
+    $cmp = bccomp($total, $paid, 5);
+    if ($cmp == 0 || $cmp == -1) { // If paid amount is greater or equal to total amount
+      $statusId = CRM_Core_OptionGroup::getValue('contribution_status', 'Completed', 'name');
+      CRM_Core_DAO::setFieldValue('CRM_Contribute_BAO_Contribution', $params['contribution_id'], 'contribution_status_id', $statusId);
```

Comment:
All my other comments have been stylistic about api conventions.

However, this one is important for functional reasons.

We need to call civicrm_api3('contribution', 'completetransaction', array('contribution_id' => $contribution['id']) 
here instead of setting the value to completed

That is the function that will take responsibility for all the flow-on effects of creating a payment - including updating memberships & participant records & sending out emails & to my mind the whole point of this api is that if a payment causes an order to be fulfilled all the associated actions will then happen